src/util/psbt/mod.rs

```diff
@@ -208,6 +217,471 @@ impl PartiallySignedTransaction {
 
         Ok(())
     }
+
+    /// Attempts to create _all_ the required signatures for this PSBT using `k`.
+    ///
+    /// If you just want to sign an input with one specific key consider using `sighash_ecdsa`. This
+    /// function does not support scripts that contain `OP_CODESEPARATOR`.
+    ///
+    /// # Panics
+    /// If an attempt is made to sign a taproot input.
+    pub fn sign<C: Signing>(&mut self, k: impl GetKey + Copy, secp: &Secp256k1<C>) -> Result<(), SignError> {
+        let tx = self.unsigned_tx.clone();
+        let mut cache = SighashCache::new(&tx);
+
+        for i in 0..self.inputs.len() {
+            match self.signing_algorithm(i).expect("i is valid") {
+                SigningAlgorithm::Ecdsa => self.bip32_sign_ecdsa(k, i, &mut cache, secp)?,
+                SigningAlgorithm::Schnorr => unimplemented!("taproot PSBT signing not yet implemented"), 
+            }
+        }
+        Ok(())
+    }
+
+    /// Attempts to create all signatures required by this PSBT's `bip32_derivation` field, adding
+    /// them to `partial_sigs`.
+    ///
+    /// ## Returns
+    /// If an error is returned some signatures may already have been added to the PSBT. Since
+    /// `partial_sigs` is a [`BTreeMap`] it is safe to retry, previous sigs will be overwritten.
+    fn bip32_sign_ecdsa<C: Signing, T: Deref<Target=Transaction>>(
+        &mut self,
+        k: impl GetKey,
+        input_index: usize,
+        cache: &mut SighashCache<T>,
+        secp: &Secp256k1<C>
+    ) -> Result<(), SignError> {
+        let (msg, sighash_ty) = self.sighash_ecdsa(input_index, cache)?;
+        let input = &self.inputs[input_index]; // index checked in call to `sighash_ecdsa`;
+        let mut sigs = Vec::with_capacity(input.bip32_derivation.len());
```

Comment:
Can we create a BtreeMap here directly? It would be more efficient to use it instead of putting sigs into a vector and then inserting them into a map

Reply:
Got one step better, I just inserted into the `partial_sigs` map during the loop!